### PR TITLE
Add multi-proxy support to load gen

### PR DESCRIPTION
**Pull request by arun-koshy**
Allows us to scale load gen by executing a distributed load across multiple fullnode proxies. Throughput is still limited because of various RPC errors at high TPS. Changes to the `SuiClient` sdk coming in future PRs

Example usage - 
```
cargo run --package sui-benchmark --bin stress -- --log-path /tmp/stress.log --num-client-threads 24 --num-server-threads 1 --num-transfer-accounts 2 bench --target-qps 100 --num-workers 24 --in-flight-ratio 30  --transfer-object 50 --shared-counter 25 --delegation 25 --local false --stress-stat-collection --genesis-blob-path ~/labnet-genesis.blob --keystore-path ~/labnet-sui.keystore --primary-gas-id 0x59931dcac57ba20d75321acaf55e8eb5a2c47e9f --fullnode-rpc-addresses=http://lax-suifn-26e60.labnet.sui.io:9000,http://ewr-suifn-26c2a.labnet.sui.io:9000,http://lax-suifn-573b2.labnet.sui.io:9000,http://lax-suifn-1d4f1.labnet.sui.io:9000 --use-fullnode-for-execution true --use-fullnode-for-reconfig true


Benchmark Report:
+-------------+-----+--------+---------------+---------------+---------------+
| duration(s) | tps | error% | latency (min) | latency (p50) | latency (p99) |
+============================================================================+
| 67          | 83  | 0      | 153           | 289           | 8767          |
+-------------+-----+--------+---------------+---------------+---------------+
Stress Performance Report:
+-----------+-----+-----+
| metric    | p50 | p99 |
+=======================+
| cpu usage | 4   | 58  |
+-----------+-----+-----+
```

### CODE REVIEW

File: crates/sui-benchmark/src/benchmark_setup.rs [hunk 1]
```diff
@@ -194,76 +193,84 @@ impl Env {
...
+                fullnodes.push(Arc::new(FullNodeProxy::from_url(&fullnode_rpc_url).await?));
+            }
+            fullnodes
         } else {
             info!("Using LocalValidatorAggregatorProxy");
             let reconfig_fullnode_rpc_url =
                 if use_fullnode_for_reconfig {
-                    Some(fullnode_rpc_url.expect(
+                    // Only need to use one full node for reconfiguration.
+                    Some(fullnode_rpc_urls.get(0).expect(
```
Comment (sadhansood) [inline crates/sui-benchmark/src/benchmark_setup.rs:213]: Maybe use a random one?

File: crates/sui-benchmark/src/benchmark_setup.rs [hunk 2]
```diff
@@ -194,76 +193,84 @@ impl Env {
...
+                    reconfig_fullnode_rpc_url.map(|x| &**x),
                 )
                 .await,
-            )
+            )]
         };
         info!(
             "Reconfiguration - Reconfiguration to epoch {} is done",
-            proxy.get_current_epoch(),
+            proxies[0].get_current_epoch(),
```
Comment (sadhansood) [inline crates/sui-benchmark/src/benchmark_setup.rs:232]: use a random proxy here too?

File: crates/sui-benchmark/src/benchmark_setup.rs [hunk 3]
```diff
@@ -194,76 +193,84 @@ impl Env {
...
-                keypair.clone(),
-            ),
-            pay_coin: (pay_coin.compute_object_reference(), pay_coin.owner, keypair),
-            pay_coin_type_tag: ttag,
-            validator_proxy: proxy,
-        })
+        let mut proxy_gas_and_coins = vec![];
+
+        for proxy in proxies.iter() {
+            let offset = ObjectID::from_hex_literal(primary_gas_id)?;
```
Comment (sadhansood) [inline crates/sui-benchmark/src/benchmark_setup.rs:238]: It's unclear to me why we need to go through all proxies to generate gas. Any one should be enough, no?

File: crates/sui-benchmark/src/bin/stress.rs [hunk 4]
```diff
@@ -67,10 +67,12 @@ async fn main() -> Result<()> {
...
     let cloned_barrier = barrier.clone();
     let env = if opts.local { Env::Local } else { Env::Remote };
-    let benchmark_setup = env.setup(cloned_barrier, &registry, &opts).await?;
+    let proxy_gas_and_coins = env.setup(cloned_barrier, &registry, &opts).await?;
     let system_state_observer = {
+        // Only need to get system state from one proxy as it is shared for the
+        // whole network.
         let mut system_state_observer =
-            SystemStateObserver::new(benchmark_setup.validator_proxy.clone());
+            SystemStateObserver::new(proxy_gas_and_coins[0].proxy.clone());
```
Comment (sadhansood) [inline crates/sui-benchmark/src/bin/stress.rs:75]: random proxy here too.

File: crates/sui-benchmark/src/drivers/bench_driver.rs [hunk 5]
```diff
@@ -241,8 +242,7 @@ async fn ctrl_c() -> std::io::Result<()> {
 impl Driver<(BenchmarkStats, StressStats)> for BenchDriver {
     async fn run(
         &self,
-        workloads: Vec<WorkloadInfo>,
-        proxy: Arc<dyn ValidatorProxy + Sync + Send>,
+        proxy_workloads: Vec<(Arc<dyn ValidatorProxy + Send + Sync>, Vec<WorkloadInfo>)>,
```
Comment (sadhansood) [inline crates/sui-benchmark/src/drivers/bench_driver.rs:245]: @arun-koshy - It seems a much easier approach could be to have a proxy which can round robin to several full nodes behind the scene. I don't think we need to have a workload per full node in that scenario.

**Review by longbowlu — COMMENTED** [review]
overall the logic looks good to me. Btw this seems like a good PR if we can extract some code into their own functions.

File: crates/sui-benchmark/src/benchmark_setup.rs (hunk 3 above)
Comment (arun-koshy) [reply crates/sui-benchmark/src/benchmark_setup.rs:238]: After this week, looks like scaling this is not needed anymore urgently. But the idea was that if we use a separate gas coin for each proxy we can then both generate gas coins in parallel for each proxy  without contention over primary gas id (if its a bottleneck) and we can ensure that we don't have any issues splitting gas coins if the load + num of proxies is larger than the total amount of one shared primary gas/pay coin.

File: crates/sui-benchmark/src/drivers/bench_driver.rs (hunk 5 above)
Comment (arun-koshy) [reply crates/sui-benchmark/src/drivers/bench_driver.rs:245]: Again we don't look like we will need scaling with fullnodes for the initial benchmarking, but without separating the workload we will have to wait for local execution at each full node to execute the subsequent transactions which would limit our throughput.

Comment (arun-koshy) [reply crates/sui-benchmark/src/drivers/bench_driver.rs:245]: Again we don't look like we will need scaling with fullnodes for the initial benchmarking, but without separating the workload we will have to wait for local execution at each full node to execute the subsequent transactions which would limit our throughput.

**Review by sadhansood — APPROVED** [review]

**Pull request merged**